blocks/api/serializer.js

```diff
@@ -36,35 +36,42 @@ export function getSaveContent( save, attributes ) {
 	return wp.element.renderToString( rawContent );
 }
 
+const escapeDoubleQuotes = value => 'string' === typeof value
+	? value.replace( /"/g, '\"' )
+	: value;
+
 /**
- * Returns comment attributes as serialized string, determined by subset of
- * difference between actual attributes of a block and those expected based
- * on its settings.
+ * Returns attributes which ought to be saved
+ * and serialized into the block comment header
+ *
+ * When a block exists in memory it contains as its attributes
+ * both those which come from the block comment header _and_
+ * those which come from parsing the contents of the block.
+ * Additionally they may live in a form fine for memory but
+ * which isn't valid for serialization.
+ *
+ * This function returns a filtered set of attributes which are
+ * the ones which need to be saved in order to ensure a full
+ * serialization and they are in a form which is safe to store.
  *
- * @param  {Object} realAttributes     Actual block attributes
- * @param  {Object} expectedAttributes Expected block attributes
- * @return {string}                    Comment attributes
+ * @param {Object<String,*>}   allAttributes         Attributes from in-memory block data
+ * @param {Object<String,*>}   attributesFromContent Attributes which are inferred from block content
+ * @returns {Object<String,*>} filtered set of attributes for minimum safe save/serialization
  */
-export function getCommentAttributes( realAttributes, expectedAttributes ) {
-	// Find difference and build into object subset of attributes.
-	const keys = difference(
-		Object.keys( realAttributes ),
-		Object.keys( expectedAttributes )
-	);
-
-	// Serialize the comment attributes as `key="value"`.
-	return keys.reduce( ( memo, key ) => {
-		const value = realAttributes[ key ];
-		if ( undefined === value ) {
-			return memo;
-		}
+export function getCommentAttributes( allAttributes, attributesFromContent ) {
+	// Iterate over attributes and produce the set to save
+	return reduce(
+		Object.keys( allAttributes ),
+		( toSave, key ) => {
+			const allValue = allAttributes[ key ];
+			const contentValue = attributesFromContent[ key ];
 
-		if ( 'string' === typeof value ) {
-			return memo + `${ key }="${ value.replace( '"', '\"' ) }" `;
-		}
-
-		return memo + `${ key }="${ value }" `;
-	}, '' );
+			return ! ( contentValue !== undefined || allValue === undefined )
+				? Object.assign( toSave, { [ key ]: escapeDoubleQuotes( allValue ) } )
```

Comment:
It feels a bit weird to me that we're encoding the attributes values in this function but we generate the `key="value"` in the other function. Serialization of the comment attributes is split between two methods.

I'd think we should avoid encoding here, or generate the complete string here because encoding the values without generating a string has no sense to me.